### ISSUE TITLE
WebKitWebProcess crashes because BadValue X11 Error, with NVidia drivers

### ISSUE THREAD

**Issue by Cybolic** [tiliado/nuvolaruntime#342]
I get an error message stating "Integration error  Failed to inject JavaScript API. Main object not found.' The application might not function properly." I'm running the latest flatpak version installed with --system.

![screenshot-20170511170829-1370x739](https://cloud.githubusercontent.com/assets/1196197/25953708/d2b2ba5c-366c-11e7-997a-c8413cb8ad8a.png)

**Comment by fenryxo** [comment]
Hello @Cybolic. Thanks for taking the time to report your issue. I need more information to investigate it further:

* Which [distribution and version](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines#your-system) do you use?
* The debugging output from a terminal as specified in the [Bug Reporting Guidelines](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines#debugging-output).
* Is the issue always reproducible or does it happen randomly?

**Comment by Cybolic** [comment]
Sorry about that.
My system is Antergos Arch, updated today.
Yes, the issue happens constantly. The popup will only appear after pressing reload, before that, the application just loads to a blank grey screen (in the area where the Google Play Music logo is visible in the screenshot above).
[nuvola-gogle-play-music-log.txt](https://github.com/tiliado/nuvolaplayer/files/993498/nuvola-gogle-play-music-log.txt)

**Comment by fenryxo** [comment]
Thanks for the logs. The following snippet might be the cause of the issue.

```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast

(WebKitWebProcess:18): Gdk-ERROR **: The program 'WebKitWebProcess' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadValue (integer parameter out of range for operation)'.
  (Details: serial 216 error_code 2 request_code 154 (GLX) minor_code 34)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

Is this a clean Nuvola flatpak installation or have you upgraded a previous flatpak build?

**Comment by Cybolic** [comment]
This is an upgraded installation (flatpak update --system).

**Comment by Cybolic** [comment]
Purging the flatpak installation and reinstalling didn't change anything.

**Comment by jamiehutber** [comment]
I'm always having the same or similar issue.

I get:

`)]}'

[["er",null,null,null,null,405,null,null,null,9]
]`

when I enter the app.

**Comment by fenryxo** [comment]
Hello @jamiehutber. That seems to be a stale sign-up page (a funny Google's bug). Pressing home button should help. If not, please create a new ticket for your issue.

**Comment by jamiehutber** [comment]
Ah yes, thanks. I do still get an error when I get through to the login page, once entering a password it just goes black. Could I ask where the log files are stored? Then i will create a new ticket.

**Comment by fenryxo** [comment]
See [Bug Reporting Guidelines](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines#debugging-output).

**Comment by fenryxo** [comment]
@Cybolic 

* Have you upgraded any of packages related to graphics drivers recently?
* I'm going to upgrade WebKitGTK (there have been a few fixes recently) in the flatpak builds and we'll look at this issue again.

**Comment by Cybolic** [comment]
Yes, my NVIDIA driver was upgraded to the latest version (381.22) in the process of updating the system.

**Comment by fenryxo** [comment]
Hello @Cybolic. I've updated WebKitGTK+ to 2.16.2. Could your ru `flatpak update --system` and `flatpak update --user` and then check whether the issue still occurs?

If yes, is it reproducible in WebKitGTK+ MiniBrowser?

```
flatpak run \
--command=/app/libexec/webkit2gtk-4.0/MiniBrowser \
eu.tiliado.NuvolaAppGooglePlayMusic//stable \
https://play.google.com/music
```

**Comment by excieve** [comment]
FYI: I was having this issue when installing with the GNOME Software button on Fedora but apparently it doesn't bring in some dependencies, such as GNOME platform and video driver libs. Installing with the command line works fine.

**Comment by fenryxo** [comment]
Thanks @excieve. That's interesting. Could compare the output of following command in case of installation with GNOME software and commandline?

```
for KIND in app runtime; do
for LOCATION in user system; do
echo; echo ${LOCATION}-${KIND}:
flatpak list --$LOCATION --$KIND --show-details -v
done; done
```

**Comment by Cybolic** [comment]
Yup! After updating the player works again. Feel free to close this :)

**Comment by fenryxo** [comment]
Ok @Cybolic. Glad to hear Nuvola works again.

@excieve, could it be possible that GNOME Platform received an update between your GNOME Software and command-line way of installation?

**Comment by excieve** [comment]
@fenryxo Sure, that's possible. I'll experiment with these two types of installation a bit later and will report if there's something unusual.

**Comment by vanyasem** [comment]
I have the same issue. Clean flatpack installation, Elementary OS. I've updated NVidia drivers recently. Same logs. I see there's a new version of WebKitGTK+ available, so updating it may fix the issue.

**Comment by vanyasem** [comment]
None of the music services work, white screen after clicking any button (i.e. login).

**Comment by vanyasem** [comment]
[log.txt](https://github.com/tiliado/nuvolaruntime/files/1047378/log.txt)
That's my Debug log with Deezer.

**Comment by vanyasem** [comment]
And no, the issue is not reproducible in WebKitGtk MiniBrowser.

**Comment by vanyasem** [comment]
Also, you should probably change the title to "WebKitWebProcess crashes because of BadValue X11 Error". Because this issue affects all music services, not just Google Music (Tested GM, Deezer, and Yandex Music myself).

**Comment by fenryxo** [comment]
> there's a new version of WebKitGTK+ available

Nuvola flatpaks contain the latest stable version of WebKitGTK - 2.16.3.

> That's my Debug log with Deezer.

Could you run Nuvola multiple times and check whether these two error situations always occur and in the same order or whether it is random?


Error 1:

```
(WebKitPluginProcess2:102): GLib-GObject-WARNING **: cannot register existing type 'CustomSignalsUnix'
(WebKitPluginProcess2:102): GLib-GObject-CRITICAL **: g_object_new: assertion 'G_TYPE_IS_OBJECT (object_type)' failed
(WebKitPluginProcess2:102): GLib-GObject-WARNING **: invalid (NULL) pointer instance
(WebKitPluginProcess2:102): GLib-GObject-CRITICAL **: g_signal_emit_by_name: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
```

Error 2:

```
Failed to open VDPAU backend libvdpau_nvidia.so: cannot open shared object file: No such file or directory
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast

(WebKitWebProcess:17): Gdk-ERROR **: The program 'WebKitWebProcess' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadValue (integer parameter out of range for operation)'.
  (Details: serial 215 error_code 2 request_code 154 (GLX) minor_code 34)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

**Comment by vanyasem** [comment]
[log2.txt](https://github.com/tiliado/nuvolaruntime/files/1047557/log2.txt)
[log3.txt](https://github.com/tiliado/nuvolaruntime/files/1047558/log3.txt)
[log4.txt](https://github.com/tiliado/nuvolaruntime/files/1047559/log4.txt)
3 more logs. I don't have time to look at them now, as I'm late at the metting.

**Comment by vanyasem** [comment]
Actually, yes, they seem to appear in the same order even on same lines.

**Comment by fenryxo** [comment]
I must admit that I have no experience with nvidia graphic cards. So I'm going to ask a lot of questions to understand the situation.

* Do you use open source drivers of Nvidia's closed source drivers?
* How have you updated your drivers recently?

**Comment by vanyasem** [comment]
I've updated my driver's using apt update, as usual. As a package.

**Comment by vanyasem** [comment]
It's okay to ask questions, the thing is to resolve the issue as a result. I'll be able to give more details later, as I'm not at home rn.

**Comment by fenryxo** [comment]
`sudo ldconfig -p | grep -i gl.so` could also provide some info.

**Comment by vanyasem** [comment]
`	libwayland-egl.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libwayland-egl.so.1
	libwayland-egl.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libwayland-egl.so
	libobs-opengl.so.0 (libc6,x86-64) => /usr/lib/libobs-opengl.so.0
	libobs-opengl.so (libc6,x86-64) => /usr/lib/libobs-opengl.so
	libcogl.so.20 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libcogl.so.20
	libcogl.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libcogl.so
	libQtOpenGL.so.4 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libQtOpenGL.so.4
	libOpenGL.so.0 (libc6,x86-64) => /usr/lib/nvidia-375/libOpenGL.so.0
	libOpenGL.so (libc6,x86-64) => /usr/lib/nvidia-375/libOpenGL.so
	libGL.so.1 (libc6,x86-64) => /usr/lib/nvidia-375/libGL.so.1
	libGL.so.1 (libc6) => /usr/lib32/nvidia-375/libGL.so.1
	libGL.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libGL.so
	libGL.so (libc6,x86-64) => /usr/lib/nvidia-375/libGL.so
	libGL.so (libc6) => /usr/lib32/nvidia-375/libGL.so
	libEGL.so.1 (libc6,x86-64) => /usr/lib/nvidia-375/libEGL.so.1
	libEGL.so.1 (libc6) => /usr/lib32/nvidia-375/libEGL.so.1
	libEGL.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libEGL.so
	libEGL.so (libc6,x86-64) => /usr/lib/nvidia-375/libEGL.so
	libEGL.so (libc6) => /usr/lib32/nvidia-375/libEGL.so
`
grep output

**Comment by vanyasem** [comment]
I looked up and found out that I have proprietary NVidia drivers installed

**Comment by fenryxo** [comment]
Upstream issue: flatpak/flatpak#138

**Comment by fenryxo** [comment]
Could you find out the version of nvidia driver? Maybe from the version of the driver package. It should be 375-something. Flatpak seems to support following versions:

```
flatpak remote-ls gnome | grep nvidia
org.freedesktop.Platform.GL.nvidia-304-134
org.freedesktop.Platform.GL.nvidia-304-135
org.freedesktop.Platform.GL.nvidia-340-101
org.freedesktop.Platform.GL.nvidia-340-102
org.freedesktop.Platform.GL.nvidia-367-57 
org.freedesktop.Platform.GL.nvidia-370-28 
org.freedesktop.Platform.GL.nvidia-375-26 
org.freedesktop.Platform.GL.nvidia-375-39 
org.freedesktop.Platform.GL.nvidia-378-13 
org.freedesktop.Platform.GL.nvidia-381-09 
org.freedesktop.Platform.GL.nvidia-381-22 
```

**Comment by vanyasem** [comment]
375.66

Fri Jun  2 21:51:56 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.66                 Driver Version: 375.66                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1080    Off  | 0000:01:00.0      On |                  N/A |
| 13%   58C    P0    50W / 215W |   1745MiB /  8110MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+

**Comment by fenryxo** [comment]
Hello @Cybolic & @excieve. You have also been affected by this issue. Do you use proprietary nvidia drivers? If so, which version? Thanks.

**Comment by excieve** [comment]
In my case this is nvidia driver 381.22. There's a corresponding flatpak platform package pulled in automatically:
```
$ flatpak list
Ref                                                  Options               
eu.tiliado.Nuvola/x86_64/stable                      user,current          
eu.tiliado.NuvolaAppGooglePlayMusic/x86_64/stable    user,current          
org.freedesktop.Platform.GL.nvidia-381-22/i386/1.4   user,runtime          
org.freedesktop.Platform.GL.nvidia-381-22/x86_64/1.4 user,runtime          
org.gnome.Platform.Locale/x86_64/3.24                user,runtime,partial  
org.gnome.Platform/x86_64/3.24                       user,runtime          
eu.tiliado.Nuvola/x86_64/stable                      system,current        
eu.tiliado.NuvolaAppGooglePlayMusic/x86_64/stable    system,current        
org.freedesktop.Platform.GL.nvidia-381-22/i386/1.4   system,runtime        
org.freedesktop.Platform.GL.nvidia-381-22/x86_64/1.4 system,runtime        
org.gnome.Platform.Locale/x86_64/3.24                system,runtime,partial
org.gnome.Platform/x86_64/3.24                       system,runtime   
```

**Comment by Cybolic** [comment]
I'm on the same NVIDIA driver on Arch with the latest flatpak packages and I'm not seeing the issue anymore.

```
$ flatpak list
Ref                                                  Options               
eu.tiliado.Nuvola/x86_64/stable                      system,current        
eu.tiliado.NuvolaAppGooglePlayMusic/x86_64/stable    system,current        
org.freedesktop.Platform.GL.nvidia-381-22/x86_64/1.4 system,runtime        
org.freedesktop.Platform.Icontheme.Paper/x86_64/1.0  system,runtime        
org.gnome.Platform.Locale/x86_64/3.22                system,runtime,partial
org.gnome.Platform.Locale/x86_64/3.24                system,runtime,partial
org.gnome.Platform/x86_64/3.22                       system,runtime        
org.gnome.Platform/x86_64/3.24                       system,runtime        
org.gtk.Gtk3theme.Arc-Darker/x86_64/3.22             system,runtime
```

**Comment by vanyasem** [comment]
I updated my drivers to 381.22. I can confirm the issue is resolved. It seems that flatpak doesn't support 375.66, so it's not a bug of this app.

**Comment by fenryxo** [comment]
🎉 Great! Are both error messages 1 and two from https://github.com/tiliado/nuvolaruntime/issues/342#issuecomment-305760052 gone?

I leave this issue open as I think Nuvola should somehow detect that there is no matching nvidia driver and show a message to user instead of crashing web process. Could anyone provide me with output of the following command?

```
flatpak run --command=ls eu.tiliado.Nuvola//stable -FR /usr/lib/GL/
```

Btw @excieve, you have duplicate flatpaks in system and user destinations. Is there any reason for that?

**Comment by excieve** [comment]
> Could anyone provide me with output of the following command?

```
$ flatpak run --command=ls eu.tiliado.Nuvola//stable -FR /usr/lib/GL/
/usr/lib/GL/:
glvnd/	nvidia-381-22/	vulkan/

/usr/lib/GL/glvnd:
egl_vendor.d/

/usr/lib/GL/glvnd/egl_vendor.d:
10_nvidia.json@

/usr/lib/GL/nvidia-381-22:
bin/  extra/  glvnd@  lib@  manifest.json*  vulkan@

/usr/lib/GL/nvidia-381-22/bin:
apply_extra*

/usr/lib/GL/nvidia-381-22/extra:
glvnd/				libGLESv2_nvidia.so.381.22*  libglx.so.381.22*		    libnvidia-encode.so.381.22*		  libnvidia-gtk2.so.381.22*    libnvidia-ptxjitcompiler.so.381.22*  vulkan/
libEGL_nvidia.so.0@		libGLX_indirect.so.0@	     libnvcuvid.so.381.22*	    libnvidia-fatbinaryloader.so.381.22*  libnvidia-gtk3.so.381.22*    libnvidia-tls.so.381.22*
libEGL_nvidia.so.381.22*	libGLX_nvidia.so.0@	     libnvidia-cfg.so.381.22*	    libnvidia-fbc.so.381.22*		  libnvidia-ifr.so.381.22*     libnvidia-wfb.so.381.22*
libGLESv1_CM_nvidia.so.381.22*	libGLX_nvidia.so.381.22*     libnvidia-compiler.so.381.22*  libnvidia-glcore.so.381.22*		  libnvidia-ml.so.381.22*      libvdpau_nvidia.so.381.22*
libGLESv2_nvidia.so.2@		libcuda.so.381.22*	     libnvidia-eglcore.so.381.22*   libnvidia-glsi.so.381.22*		  libnvidia-opencl.so.381.22*  tls/

/usr/lib/GL/nvidia-381-22/extra/glvnd:
egl_vendor.d/

/usr/lib/GL/nvidia-381-22/extra/glvnd/egl_vendor.d:
10_nvidia.json

/usr/lib/GL/nvidia-381-22/extra/tls:
libnvidia-tls.so.381.22*

/usr/lib/GL/nvidia-381-22/extra/vulkan:
icd.d/

/usr/lib/GL/nvidia-381-22/extra/vulkan/icd.d:
nvidia_icd.json

/usr/lib/GL/vulkan:
icd.d/

/usr/lib/GL/vulkan/icd.d:
nvidia_icd.json@
```

> Btw @excieve, you have duplicate flatpaks in system and user destinations. Is there any reason for that?

Not sure, but I suspect this might be due to first trying to install using the GNOME Software and then manually. Maybe Software tries to install in system destination. I'll try to purge everything and then reinstall using one of the methods to see what it does.

**Comment by excieve** [comment]
So I just purged everything both system or user in flatpak and tried installing Nuvola Apps and Google Play app using GNOME Software. It actually installed everything in user destination unlike manual method (which in the [instructions](https://nuvola.tiliado.eu/#!nuvola!fedora!) suggests using sudo) installing system-wide.

However, this Software installation didn't pull in nvidia driver platform. So I executed `flatpak update --user` (without sudo) and it automatically installed `org.freedesktop.Platform.GL.nvidia-381-22` and `org.gnome.Platform.Locale`. It does install it when using the manual method from the beginning.

So there appear to be two issues:
1. GNOME Software installing per-user, while manual (when following instructions) is system-wide
2. GNOME Software method doesn't pull additional (but required) dependencies

**Comment by fenryxo** [comment]
What does `flatpak run --command=cat eu.tiliado.Nuvola//stable /sys/module/nvidia/version` show?

> GNOME Software installing per-user, while manual (when following instructions) is system-wide

I think both ways are ok. The issue is whether GNOME Software can **update** both locations.

> GNOME Software method doesn't pull additional (but required) dependencies

That's this bug: [Flatpak backend doesn't seem to install related refs](https://bugzilla.gnome.org/show_bug.cgi?id=778598)

Anyway, since there are [so many flatpak-related bugs in GNOME Software](https://github.com/tiliado/nuvolaruntime/wiki/GNOME-Software), I'm going to update instructions not to use GNOME Software at all. In addition, I'll remove `sudo` because it is not necessary.

**Comment by fenryxo** [comment]
Could anyone test master branch of Nuvola flatpak with these steps? Thanks.

1. Uninstall all Nuvola flatpaks
2. Uninstall nvidia flatpaks
3. Install `eu.tiliado.Nuvola//master`
4. Make sure nvidia flatpak is not installed
5. `flatpak run eu.tiliado.Nuvola//master -D`
6. Nuvola should show error dialog about missing driver and a link to [Graphics Drivers wiki page](https://github.com/tiliado/nuvolaruntime/wiki/Graphics-Drivers)
7. Install nvidia flatpak according to instructions.
8. `flatpak run eu.tiliado.Nuvola//master -D`
9. There should be no error.

**Comment by fenryxo** [comment]
Changes have been pushed to the stable flatpak branch. Not sure whether it works as I haven't received any feedback. We'll see.

**Comment by fenryxo** [comment]
Fix released in Nuvola 4.5.
